### PR TITLE
feat(server): Use new journal format

### PR DESCRIPTION
**Pull request by dranikpg**
This PR is the most basic replacement for the current pipeline.

Maybe we should move to something more advanced already here... or just make sure the basics work alright. Not sure

### CODE REVIEW

File: src/server/journal/executor.cc
Comment (dranikpg) [inline src/server/journal/executor.cc:15]: I'm not sure if it deserves a whole separate file and what to do with it in the future.

I recently introduced this cut-down serializer_commons. But in the subsequent PR we should probably unite all the parsing stuff further

Comment (dranikpg) [reply src/server/journal/executor.cc:15]: Besides this class will grow in the future and become more complicated to coordinate on transactions

File: src/server/journal/executor.cc
Comment (dranikpg) [inline src/server/journal/executor.cc:30]: We need to somehow get the error back from the command.

We do now want it to write reponses for optimal performance. I guess we can either:
- Replace it with a custom sink that looks for the RESP error sequence and reports it
- Add an overloaded serializer (they are virtual, so it should be possible) that reports only errors.

The second approach seems to be better

File: src/server/journal/journal.cc
Comment (dranikpg) [inline src/server/journal/journal.cc:96]: Besides, I haven't checked any of the file journal code. Its not actively used, so I guess we can leave it half-baked until future

File: src/server/rdb_load.cc
Comment (dranikpg) [inline src/server/rdb_load.cc:1607]: I don't like the fact we're implicitly taking the script manager from the service, but maybe its better than passing both?

File: src/server/rdb_load.cc
Comment (dranikpg) [inline src/server/rdb_load.cc:1764]: This is important. This way we make sure the journal entries executed only after their corresponding rdb has been parsed and inserted

File: src/server/rdb_load.cc
Comment (dranikpg) [inline src/server/rdb_load.cc:1939]: Once we merge parsing utils, we can add an explicit EOF error. This way we don't need to know the number of entries.

File: src/server/snapshot.cc
Comment (dranikpg) [inline src/server/snapshot.cc:285]: Some notes about the journal change listener in snapshot. We discussed both of them, but after a second thought 😅... we can't actually execute the journal after we've written the original value. See my example

File: src/server/snapshot.cc
Comment (dranikpg) [inline src/server/snapshot.cc:304]: This is also an issue we discussed - added a comment here

File: src/server/transaction.cc
Comment (dranikpg) [inline src/server/transaction.cc:1215]: Currently, we don't have commands with tail arguments (arguments after we've listed all per-shard values). Lets see how they'll be implemented and according to it we'll do it here.

File: src/server/transaction.cc
Comment (dranikpg) [inline src/server/transaction.cc:1209]: Maybe we should also skip blocking, because I didn't test them properly. Not sure they'll work out of the box

File: src/server/main_service.cc
Comment (dranikpg) [inline src/server/main_service.cc:551]: Forgot to comment this. During full sync replica is in loading state and it doesn't allow executing commands. We change it so it allows handling commands originating from a journal executors

File: src/server/snapshot.cc
Comment (dranikpg) [reply src/server/snapshot.cc:285]: The second bug should be impossible in theory, If I'm not wrong

File: src/server/transaction.h
Comment (adiholden) [inline src/server/transaction.h:289]: I usually go with the approach of giving more specific variable/members names and not use comment.
Here it can be cmd_with_args_

File: src/server/transaction.cc
Comment (adiholden) [inline src/server/transaction.cc:1221]: can we write 
if (unique_shard_cnt_ == 1 ||  args_.empty()) 
entry_payload = full_args_;
else 
entry_payload =
        make_pair(facade::ToSV(full_args_.front()), ShardArgsInShard(shard->shard_id()));

?

File: src/server/transaction.cc
Comment (adiholden) [inline src/server/transaction.cc:1205]: There is no flow now in the code in which shard = null right?

File: src/server/snapshot.cc
Comment (adiholden) [inline src/server/snapshot.cc:278]: But the on db change should be called before the change.
So in this example the on db change will wirte v1,v2 to the rdb

File: src/server/snapshot.cc
Comment (adiholden) [inline src/server/snapshot.cc:282]: yes we force flash before moving to the next db index

File: src/server/transaction.cc
Comment (dranikpg) [reply src/server/transaction.cc:1221]: I would have to change ShardArgInShard to return an empty slice. Currently there is a `DCHECK(!args_.empty())` in ShardArgInShard.

I think it would be safe to change the mechanics, but it would no longer guard the other transaction code from bugs (incorrect shard access)

File: src/server/transaction.cc
Comment (dranikpg) [reply src/server/transaction.cc:1205]: Yes

File: src/server/snapshot.cc
Comment (dranikpg) [reply src/server/snapshot.cc:278]: Ah, yes, it will have a higher snapshot version 😅 So it should be fine in all cases, I guess

File: src/server/server_family.cc
Comment (adiholden) [inline src/server/server_family.cc:573]: I think this is part of the next commit right?

File: src/server/rdb_save.cc
Comment (adiholden) [inline src/server/rdb_save.cc:694]: I see only one place calling this method with 1 entry. Why not pass const& journal::Entry ?

File: src/server/server_family.cc
Comment (dranikpg) [reply src/server/server_family.cc:573]: Yes, please ignore this one line 🙂  I will remove it when I add fixes to this

File: src/server/rdb_save.cc
Comment (dranikpg) [reply src/server/rdb_save.cc:694]: In the next PR the journal listener in dflycmd will be buffered, so it will be (according to current plans, but not sure) used there.

I can remove it, but it'll be lost efforts

File: src/server/transaction.cc
Comment (adiholden) [reply src/server/transaction.cc:1221]: I saw the check inside ShardArgInShard and I think its better to keep it there for now.
My suggestion was not to move the args_.empty() case to the initialisation using ShardArgsInShard but to the one using full_args_

**Review by adiholden — APPROVED** [review]

File: src/server/transaction.cc
Comment (dranikpg) [reply src/server/transaction.cc:1221]: Indeed

File: src/server/transaction.h
Comment (dranikpg) [reply src/server/transaction.h:289]: I called it `cmd_with_full_args_`, but kept a comment is its more clear

**Comment by dranikpg** [comment]
I've commited fixes to the comments above + the replica part

File: src/server/rdb_load.cc
Comment (adiholden) [inline src/server/rdb_load.cc:1929]: In this flow JournalReader is created for each journal entry.
JournalReader has a private member buf_ which grows as you read the command, I think you have many allocations in this flow

**Review by adiholden — DISMISSED** [review]

File: src/server/rdb_load.cc
Comment (dranikpg) [reply src/server/rdb_load.cc:1929]: Indeed... I haven't thought about this at all - just wanted to get it working

But instead of having of adding a member `JournalReader`, I would rather make it re-use the `mem_buffer_` from rdb loader

Comment (dranikpg) [reply src/server/rdb_load.cc:1929]: This would also make it possible to precisely control the lifetime of the entries in the future

**Comment by dranikpg** [comment]
I've added `JournalReader` as a member because re-using the buffer is not directly possible without some hacks. 

I've also found a  cancellation bug, I'm trying to resolve it 😢

File: src/server/common.h
Comment (adiholden) [inline src/server/common.h:271]: can you explain why try lock?

File: src/server/replica.cc
Comment (adiholden) [inline src/server/replica.cc]: Why it did not work?

File: src/server/rdb_load.cc
Comment (adiholden) [inline src/server/rdb_load.cc]: what about the dbid? How is this working now?

Comment (dranikpg) [reply src/server/rdb_load.cc]: Indeed 😅

File: src/server/common.h
Comment (dranikpg) [reply src/server/common.h:271]: Because it happened so that `Error()` was called from inside the error handler. We should maybe allow this in the future, this is why I've added a TODO

File: src/server/rdb_load.cc
Comment (dranikpg) [reply src/server/rdb_load.cc]: This should be catchable by tests once we have the new testing data generator

File: src/server/journal/serializer.h
Comment (adiholden) [inline src/server/journal/serializer.h:49]: I guess we are not using now the logic of the db index in the constructor, if so it can be dropped.

**Review by adiholden — APPROVED** [review]

**Pull request merged**
